python/dolfinx/nls/petsc.py

```diff
@@ -72,3 +98,434 @@ def setP(self, P: types.FunctionType, Pmat: PETSc.Mat):  # type: ignore
 
         """
         super().setP(P, Pmat)
+
+
+def _create_snes_matrices_and_vectors(
+    a: typing.Union[list[list[fem.Form]], fem.Form],
+    L: typing.Union[list[fem.Form], fem.Form],
+    P: typing.Union[list[list[fem.Form]], list[fem.Form], fem.Form, None],
+    assembly_type: fem.petsc.AssemblyType,
```

Comment:
I think `assembly_type` should be a PETSc Mat/Vec type. This would allow the caller to not only specify 'nest`, but also other PETSc matrix types.

Reply:
Already discussed this, it was also my preference, but @jorgensd told me it's not possible to discern between nest and block types. I do see `MATNEST` and `MATBLOCK` though: https://petsc.org/release/manualpages/Mat/MatType/ - would these work @jorgensd ?

Reply:
We don't use `createBlockMat` (https://petsc.org/release/manualpages/Mat/MatCreateBlockMat/).
We use `createMat` and attach an index set to it based on the concatenated index maps.

Reply:
We don't need to discern between block and nest types. If the forms are a list, then it's block or nest. Nest is used if the user asks for nest, otherwise `MATAIJ` or whatever else they want.

The interface should probably have options for the vector and the matrix (matrices for nest), since they can differ. Even for a single matrix or a matrix in a nest there are options, e.g. block CSR. See https://github.com/FEniCS/dolfinx/blob/192222e278fb2809abe97c3d46ecda534fb6bcdf/python/dolfinx/fem/petsc.py#L174 where the users passes the `MatType`. The other "create_matrix functions should have that option too.

Reply:
Focusing on the matrix, the user should do one of the following:
1. Send in `None` for the matrix type. In this case the default PETSc matrix type will be used. If the forms are blocked, the assembly will be blocked; or
2. Send in `MATNEST` as the matrix type. In this case each block will have the default matrix type; or
3. Send in a list-of-lists with the matrix type for each block in a `MATNEST` matrix. 

The C++ interfaces for creating the matrices support these options. The code should handle a PETSc prefix to support setting of matrix types from the command line.